File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        visited = {}
        while headA != None:
            visited[headA] = True
            headA = headA.next
        
        while headB != None:
            if visited.get(headB, False):
                return headB
            headB = headB.next
        return None    
"""

# O(1) Space
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        a_start = headA
        b_start = headB
        
        while headA != headB:
            if headA is None:
                headA = b_start
                continue
            if headB is None:
                headB = a_start
                continue
            headA = headA.next
            headB = headB.next
        
        return headA

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/*
func getIntersectionNode(headA, headB *ListNode) *ListNode {
    visited := map[*ListNode]int{}
    
    cur1 := headA
    cur2 := headB
    for cur1 != nil || cur2 != nil {
        if cur1 != nil {
            if _, ok := visited[cur1]; ok {
                return cur1
            } 
            visited[cur1]++
            cur1 = cur1.Next
        }
        
        if cur2 != nil {
            if _, ok := visited[cur2]; ok {
                return cur2
            }
            visited[cur2]++
            cur2 = cur2.Next
        }
    }
    
    return nil
}
*/
func getIntersectionNode(headA, headB *ListNode) *ListNode {
    
    cur1 := headA
    cur2 := headB
    
    for cur1 != cur2 {
        if cur1 == nil {
            cur1 = headB
        } else {
            cur1 = cur1.Next
        }
        
        if cur2 == nil {
            cur2 = headA
        } else {
            cur2 = cur2.Next
        }
    }
    
    return cur1
}